src/eos-updater-poll-common.c

```diff
@@ -0,0 +1,910 @@
+  g_autofree gchar *upgrade_ref = NULL;
+  gboolean on_hold = FALSE;
+
+  g_return_val_if_fail (EOS_IS_BRANCH_FILE (branch_file), FALSE);
+  g_return_val_if_fail (upgrade_refspec != NULL, FALSE);
+  g_return_val_if_fail (original_refspec != NULL, FALSE);
+
+  booted_deployment = eos_updater_get_booted_deployment (error);
+
+  if (!get_origin_refspec (booted_deployment, &booted_refspec, error))
+    return FALSE;
+
+  if (!ostree_parse_refspec (booted_refspec, &booted_remote, &booted_ref, error))
+    return FALSE;
+
+  hw_descriptors = get_hw_descriptors ();
+  vendor = cleanstr (g_strdup (g_hash_table_lookup (hw_descriptors, VENDOR_KEY)));
+  product = cleanstr (g_strdup (g_hash_table_lookup (hw_descriptors, PRODUCT_KEY)));
+
+  product_group = g_strdup_printf ("%s %s", vendor, product);
+  if (!process_branch_file (branch_file->branch_file, product_group, &on_hold, &upgrade_ref, error))
+    return FALSE;
+
+  if (on_hold)
+    upgrade_ref = g_strdup (booted_ref);
+  else
+    {
+      if (upgrade_ref == NULL)
+        {
+          message ("No product-specific branch configuration or %s found, "
+                   "following the origin file", DEFAULT_GROUP);
+          upgrade_ref = g_strdup (booted_ref);
+        }
+
+      message ("Using product branch %s", upgrade_ref);
+      *upgrade_refspec = g_strdup_printf ("%s:%s", booted_remote, upgrade_ref);
+      *original_refspec = g_strdup (booted_refspec);
+    }
+
+  if (metrics != NULL)
+    {
+      g_autoptr(EosMetricsInfo) info = NULL;
+
+      info = g_object_new (EOS_TYPE_METRICS_INFO, NULL);
+      info->vendor = g_steal_pointer (&vendor);
+      info->product = g_steal_pointer (&product);
+      info->ref = g_steal_pointer (&upgrade_ref);
+      info->on_hold = on_hold;
+      info->branch_file = g_object_ref (branch_file);
+      *metrics = g_steal_pointer (&info);
+    }
+
+  return TRUE;
+}
+
+static GVariant *
+get_repo_pull_options (const gchar *url_override,
+                       const gchar *ref)
+{
+  g_auto(GVariantBuilder) builder;
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a{sv}"));
+  if (url_override != NULL)
+    g_variant_builder_add (&builder, "{s@v}", "override-url",
+                           g_variant_new_variant (g_variant_new_string (url_override)));
+  g_variant_builder_add (&builder, "{s@v}", "flags",
+                         g_variant_new_variant (g_variant_new_int32 (OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY)));
+  g_variant_builder_add (&builder, "{s@v}", "refs",
+                         g_variant_new_variant (g_variant_new_strv (&ref, 1)));
+
+  return g_variant_ref_sink (g_variant_builder_end (&builder));
+};
+
+static gboolean
+get_extensions_url (OstreeRepo *repo,
+                    const gchar *remote_name,
+                    const gchar *url_override,
+                    gchar **extensions_url,
+                    GError **error)
+{
+  g_autofree gchar *url = NULL;
+
+  if (url_override != NULL)
+    url = g_strdup (url_override);
```

Comment:
`g_strdup` will just return `NULL` if the input value is `NULL`, so you can just check if `url` is `NULL` after attempting to `strdup` it and if so, try getting it from `ostree_repo_remote_get_url`.